rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,277 @@
+//! Convert a Parquet table to a Delta table in place
+
+use crate::{
+    errors::DeltaTableError,
+    open_table,
+    operations::create::CreateBuilder,
+    protocol::{Action, Add, SaveMode},
+    schema::Schema,
+    storage::DeltaObjectStore,
+    DeltaTable, DeltaTablePartition, SchemaDataType, SchemaField,
+};
+use arrow::{datatypes::Schema as ArrowSchema, error::ArrowError};
+use futures::StreamExt;
+use log::*;
+use object_store::{ObjectMeta, ObjectStore};
+use parquet::{
+    arrow::async_reader::{ParquetObjectReader, ParquetRecordBatchStreamBuilder},
+    errors::ParquetError,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    num::TryFromIntError,
+    sync::Arc,
+};
+use url::Url;
+
+/// Error converting a Parquet table to a Delta table
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Object store error
+    #[error("Object store error: {0}")]
+    ObjectStore(#[from] object_store::Error),
+    /// Arrow error
+    #[error("Arrow error: {0}")]
+    Arrow(#[from] ArrowError),
+    /// Parquet error
+    #[error("Parquet error: {0}")]
+    Parquet(#[from] ParquetError),
+    /// DeltaTable error
+    #[error("DeltaTable error: {0}")]
+    DeltaTable(#[from] DeltaTableError),
+    /// Error converting usize to i64
+    #[error("Error converting usize to i64: {0}")]
+    TryFromUsize(#[from] TryFromIntError),
+    /// Error canonicalizing the given path
+    #[error("Error canonicalizing the given path: {0}")]
+    CanonicalizePath(#[from] std::io::Error),
+    /// Error converting the path into an URL
+    #[error("Error converting the path into an URL")]
+    UrlFromFilePath,
+}
+
+/// Convert a Parquet table to a Delta table in place
+pub async fn convert_to_delta(path: &str) -> Result<DeltaTable, Error> {
```

Comment:
I would recommend this API taking an already configured `ObjectStore`. The object store can have a number of configured parameters that exist outside of the `path` We have plenty of other places where URL and path parsing exist and have caused us enough grief.

Reply:
Thanks for the recommendation. I updated the function to take an `ObjectStoreRef` as argument. I also added a `convert_to_delta_from_path` function that takes a path as argument, since [the `convertToDelta` command](https://docs.delta.io/latest/api/python/index.html#delta.tables.DeltaTable.convertToDelta) supports path. Would it make sense to support both functions in the API?